kafka consumer may read previous message if commit message offset failed

```
[2024/07/03 23:24:56.556 +08:00] [ERROR] [consumer.go:136] ["commit message failed, just continue"] [error="Broker: Unknown member"]
[2024/07/03 23:24:56.587 +08:00] [ERROR] [consumer.go:126] ["read message failed, just continue to retry"] [error="Application maximum poll interval (300000ms) exceeded by 221ms"]
[2024/07/03 23:24:56.748 +08:00] [PANIC] [writer.go:388] ["RowChangedEvent fallback row, ignore it"] [commitTs=450892761914671327] [offset=55842905] [watermark=450892849772232846] [watermarkOffset=55934110] [partition=2] [tableID=339] [schema=sysbench_64_7000w] [table=sbtest57] [stack="main.(*writer).WriteMessage\n\[tgithub.com/pingcap/tiflow/cmd/kafka-consumer/writer.go:388\nmain.(*consumer).Consume\n\tgithub.com/pingcap/tiflow/cmd/kafka-consumer/consumer.go:129\nmain.main.func2\n\tgithub.com/pingcap/tiflow/cmd/kafka-consumer/main.go:101](http://tgithub.com/pingcap/tiflow/cmd/kafka-consumer/writer.go:388/nmain.(*consumer).Consume/n/tgithub.com/pingcap/tiflow/cmd/kafka-consumer/consumer.go:129/nmain.main.func2/n/tgithub.com/pingcap/tiflow/cmd/kafka-consumer/main.go:101)"]
```

--- a/cmd/kafka-consumer/writer.go
+++ b/cmd/kafka-consumer/writer.go
@@ -388,7 +388,7 @@ func (w *writer) WriteMessage(ctx context.Context, message *kafka.Message) bool 
 				// if commit message failed, the consumer may read previous message,
 				// just ignore this message should be fine, otherwise panic.
 				if message.TopicPartition.Offset > progress.watermarkOffset {
-					log.Panic("RowChangedEvent fallback row, ignore it",
+					log.Panic("RowChangedEvent fallback row",
 						zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", message.TopicPartition.Offset),
 						zap.Uint64("watermark", watermark), zap.Any("watermarkOffset", progress.watermarkOffset),
 						zap.Int32("partition", partition), zap.Int64("tableID", tableID),

--- a/cmd/kafka-consumer/writer.go
+++ b/cmd/kafka-consumer/writer.go
@@ -384,13 +384,22 @@ func (w *writer) WriteMessage(ctx context.Context, message *kafka.Message) bool 
 			watermark := atomic.LoadUint64(&progress.watermark)
 			// if the kafka cluster is normal, this should not hit.
 			// else if the cluster is abnormal, the consumer may consume old message, then cause the watermark fallback.
-			if row.CommitTs < watermark {
-				log.Panic("RowChangedEvent fallback row, ignore it",
+			if row.CommitTs < watermark && message.TopicPartition.Offset > progress.watermarkOffset {
+				if message.TopicPartition.Offset > progress.watermarkOffset {
+					log.Panic("RowChangedEvent fallback row, ignore it",
+						zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", message.TopicPartition.Offset),
+						zap.Uint64("watermark", watermark), zap.Any("watermarkOffset", progress.watermarkOffset),
+						zap.Int32("partition", partition), zap.Int64("tableID", tableID),
+						zap.String("schema", row.TableInfo.GetSchemaName()),
+						zap.String("table", row.TableInfo.GetTableName()))
+				}
+				log.Warn("Row changed event fall back, ignore it, since consumer read old offset message",
 					zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", message.TopicPartition.Offset),
 					zap.Uint64("watermark", watermark), zap.Any("watermarkOffset", progress.watermarkOffset),
 					zap.Int32("partition", partition), zap.Int64("tableID", tableID),
 					zap.String("schema", row.TableInfo.GetSchemaName()),
 					zap.String("table", row.TableInfo.GetTableName()))
+				continue
 			}
 			group, ok := eventGroup[tableID]
 			if !ok {

--- a/cmd/kafka-consumer/writer.go
+++ b/cmd/kafka-consumer/writer.go
@@ -384,7 +384,7 @@ func (w *writer) WriteMessage(ctx context.Context, message *kafka.Message) bool 
 			watermark := atomic.LoadUint64(&progress.watermark)
 			// if the kafka cluster is normal, this should not hit.
 			// else if the cluster is abnormal, the consumer may consume old message, then cause the watermark fallback.
-			if row.CommitTs < watermark && message.TopicPartition.Offset > progress.watermarkOffset {
+			if row.CommitTs < watermark {
 				if message.TopicPartition.Offset > progress.watermarkOffset {
 					log.Panic("RowChangedEvent fallback row, ignore it",
 						zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", message.TopicPartition.Offset),

--- a/cmd/kafka-consumer/writer.go
+++ b/cmd/kafka-consumer/writer.go
@@ -385,6 +385,8 @@ func (w *writer) WriteMessage(ctx context.Context, message *kafka.Message) bool 
 			// if the kafka cluster is normal, this should not hit.
 			// else if the cluster is abnormal, the consumer may consume old message, then cause the watermark fallback.
 			if row.CommitTs < watermark {
+				// if commit message failed, the consumer may read previous message,
+				// just ignore this message should be fine, otherwise panic.
 				if message.TopicPartition.Offset > progress.watermarkOffset {
 					log.Panic("RowChangedEvent fallback row, ignore it",
 						zap.Uint64("commitTs", row.CommitTs), zap.Any("offset", message.TopicPartition.Offset),

